crates/iota-core/src/authority/shared_object_congestion_tracker.rs

```diff
@@ -137,51 +296,274 @@ impl SharedObjectCongestionTracker {
+                    // case AB: a free slot remains at the start and the end.
+                    // (occupied_slot.start_time > free_slot.start_time && occupied_slot.end_time
+                    // <
+                    // free_slot.end_time)
+                    //      | free_slot                             |
+                    //   => | free_slot | occupied_slot | free_slot |
+                    // case 0: the occupied slot perfectly overlaps with the free slot.
+                    // (occupied_slot.start_time == free_slot.start_time && occupied_slot.end_time
+                    // == free_slot.end_time)
+                    //      | free_slot     |
+                    //   => | occupied_slot |
+                    if occupied_slot.intersects(free_slot) {
+                        // The occupied slot must be within the free slot or the assigned slot is
+                        // not correct.
+                        assert!(
+                            occupied_slot.start_time >= free_slot.start_time
+                                && occupied_slot.end_time <= free_slot.end_time
+                        );
+                        // store the index of the old slot to remove it later.
+                        old_slot_index = Some(index);
+                        // case A: if a part of the free slot remains at the start, create a new
+                        // free slot.
+                        if occupied_slot.start_time > free_slot.start_time {
+                            new_slots.push(ExecutionSlot::new(
+                                free_slot.start_time,
+                                occupied_slot.start_time,
+                                false,
+                            ));
+                        }
+                        // case B: if a part of the free slot remains at the end, create a new free
+                        // slot.
+                        if occupied_slot.end_time < free_slot.end_time {
+                            new_slots.push(ExecutionSlot::new(
+                                occupied_slot.end_time,
+                                free_slot.end_time,
+                                false,
+                            ));
+                        }
+                        break;
+                    }
+                }
+                // remove the old slot and add the new slots.
+                let slots = self.object_execution_slots.get_mut(&obj.id).unwrap();
+                if let Some(i) = old_slot_index {
+                    slots.remove(i);
+                }
+                slots.push(occupied_slot);
+                slots.extend(new_slots);
+                slots.sort_by(|a, b| a.start_time.cmp(&b.start_time));
             }
         }
     }
 
-    // Returns the maximum cost of all objects.
-    pub fn max_cost(&self) -> u64 {
-        self.object_execution_cost
+    // Returns the maximum occupied slot end time over all shared objects.
```

Comment:
This is a public function used in another module, so I suggest turning the comment into a docstring.

```suggestion
    /// Returns the maximum occupied slot end time over all shared objects.
```